src/ducks/modules/network.js

```diff
@@ -39,16 +40,54 @@ function edgeExists(edges, edge) {
   );
 }
 
+function getNodesWithAdd(nodes, node) {
+  const id = nextId(nodes);
+  const uid = nextUid(nodes);
+  // Provided uid can override generated one, but not id
+  const nodeWithId = { uid, ...node, id };
+  return [...nodes, nodeWithId];
+}
+
+function getUpdatedNodes(nodes, updatedNode, full) {
+  const updatedNodes = nodes.map((node) => {
+    if (node.uid !== updatedNode.uid) { return node; }
+
+    if (full) {
+      return {
+        ...updatedNode,
+        id: node.id,
+        uid: node.uid,
+      };
+    }
+
+    return {
+      ...node,
+      ...updatedNode,
+      id: node.id,
+      uid: node.uid,
+    };
+  });
+  return updatedNodes;
+}
+
 export default function reducer(state = initialState, action = {}) {
   switch (action.type) {
     case ADD_NODE: {
-      const id = nextId(state.nodes);
-      const uid = nextUid(state.nodes);
-      // Provided uid can override generated one, but not id
-      const node = { uid, ...action.node, id };
       return {
         ...state,
-        nodes: [...state.nodes, node],
+        nodes: getNodesWithAdd(state.nodes, action.node),
+      };
+    }
+    case ADD_OR_UPDATE_NODE: {
+      let nodes = state.nodes;
+      if (state.nodes.filter(node => node.uid === action.node.uid).length > 0) {
```

Comment:
I'd probably use `find` for this, though we probably won't see any performance improvements for that :)

If you use the lodash variant there's a shorthand version you could use too:
`find(state.nodes, ['uid', action.node.uid])`

Reply:
Sounds good.